src/background.js

```diff
@@ -451,33 +448,14 @@ const rollout = {
         }
       }
     });
-  },
 
-  async main() {
-    // Listen to the captive portal when it unlocks
-    browser.captivePortal.onStateChanged.addListener(rollout.onReady);
-
-    // If the captive portal is already unlocked or doesn't exist,
-    // run the measurement
-    let captiveState = await browser.captivePortal.getState();
-    log("Captive state:", captiveState);
-    if ((captiveState === "unlocked_portal") ||
-        (captiveState === "not_captive")) {
-      await rollout.onReady({state: captiveState});
-    }
+    browser.captivePortal.onConnectivityAvailable.addListener(
+      rollout.runStartupHeuristics()
+    );
```

Comment:
Also adding only this listener is not enough. If we already have connectivity it will never fire, until we lose connectivity and get it back again.
You will need to check browser.captivePortal.getState() as well.

On every network change event we should check captive portal as well.

Reply:
This is fired when the network changes, once connectivity has been established. The propagated status is "clear".

Reply:
I suggested to Maxx that we should just not bother with network changes at all. The captive portal service takes care of establishing connectivity.
If we want to be super thorough, we'd have to check whether the user has disabled captive portal for some reason, and only in that case, check the network status as a fallback.

Reply:
(you're right that we would have to check the captive portal state on the first run though)

Reply:
The network change debouncing and the captive portal logic are likely to conflict in weird ways. This needs to be pretty thoroughly tested. Ideally what we'd do is set and save a timeout when the network changes, and after that timeout, run heuristics. If we get a captive portal "clear" or "not_captive", we'd clear the timeout and run heuristics immediately. If we get a captive portal "captive", we'd clear the timeout and wait for a "clear" or "not_captive". If we get a link "down", we'd clear the timeout and wait for link "up".

Reply:
This conflict wasn't apparent before because the captive portal logic was simply not running.

Reply:
@ddragana @nhnt11 I'm getting duplicate events on start up. Once, officially for "startup" and another when onConnectivityAvailable fires. Can I expect that onConnectivityAvailable fires once on start up?

Reply:
I looked at m.-c. code and I do not find "startup" event, but maybe I just missed it somehow. 
Anyway do not relay on that, it may be racy.

Reply:
I don't know if using the captive portal event is advisable - there are a few weird types of captive portals that we don't actually detect - and the CP web-extension API is quite new and I'm not sure we can be 100% sure there are now bugs.
Maybe we should keep the network change events as well for the time?

Reply:
Not detecting a portal is fine, IMO. We still receive the connectivity event. The detection checks the availability of a canonical URL, which is perfect for our use case no? Do we care about the portals we don't manage to detect? There's no way to handle those anyway.

I am confident that if the captive portal service is enabled, it's redundant to check both network changes and portal connectivity. We'll just end up building a complicated state machine to handle the race.

Reply:
If on the other hand you are referring to a timeout or abort of the portal check, to me that's still good indication that we shouldn't run heuristics.

Reply:
Dragana, Valentin, Maxx, and I came to a conclusion outside of this thread:

1. Check network state and captive portal state. If network up and not captive, run heuristics.
2. Add listener for network ID change. When it changes, run heuristics. Listen for captive portal state change.
    a. If we get a locked portal, wait for clear, then run heuristics again. The previous run would have been unsuccessful.
    b. If we get a clear portal, then all is well.